### DISCUSSION TITLE
libbson-dev and libmongoc-dev dependencies

### DISCUSSION THREAD

**Discussion by amdrsantos** [syslog-ng/syslog-ng#5044] [Q&A]
I have a problem because the libbson-dev and libmongoc-dev versions on the target (Debian) are older than those used to compile (on docker).
I can not update them because there is no backport for these libraries.

Is there any option to link these two statically?
Is there any option to not compile at all the syslog-ng-mod-mongodb module?

**Comment by bazsi** [comment]
The easiest is to disable mongodb support, which can be done using `--disable-mongodb` configure option while compiling from source. This is also auto-detected, as if the required libraries are not found, syslog-ng will disable mongodb automatically.

If you are compiling a debian package (either via dbld or our released tarball), it is a bit more involved. In that case, either the packaging needs to be updated so that mongodb can be optional, _or_ patch the packaging files to remove it.

**Comment by amdrsantos** [comment]
Thanks, Bazsi

I am trying to compile from sources **without changing anything**, but the result tar.gz is empty.
Here is the summary of flow that I am using:

```
wget https://github.com/syslog-ng/syslog-ng/releases/download/syslog-ng-4.3.1/syslog-ng-4.3.1.tar.gz
tar -xzf syslog-ng-4.3.1.tar.gz
cd syslog-ng-4.3.1
dbld/rules shell-debian-bookworm
```
Inside docker
```
/dbld/bootstrap
cd /dbld/build
make -j
make install
make dist
```

After "make dist", the tar.gz is empty
```
(dbld)asantos@mymachine:/dbld/build$ make dist
make --no-print-directory dist-gzip am__post_remove_distdir='@:'
make --no-print-directory distdir-am
if test -d "syslog-ng-4.3.1"; then find "syslog-ng-4.3.1" -type d ! -perm -200 -exec chmod u+w {} ';' && rm -rf "syslog-ng-4.3.1" || { sleep 5 && rm -rf "syslog-ng-4.3.1"; }; else :; fi
test -d "syslog-ng-4.3.1" || mkdir "syslog-ng-4.3.1"
 (cd lib/ivykis && make --no-print-directory top_distdir=../../syslog-ng-4.3.1 distdir=../../syslog-ng-4.3.1/lib/ivykis \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
:
test -d "../../syslog-ng-4.3.1/lib/ivykis" || mkdir "../../syslog-ng-4.3.1/lib/ivykis"
 (cd src && make  top_distdir=../../../syslog-ng-4.3.1 distdir=../../../syslog-ng-4.3.1/lib/ivykis/src \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
 (cd test && make  top_distdir=../../../syslog-ng-4.3.1 distdir=../../../syslog-ng-4.3.1/lib/ivykis/test \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
 (cd test.mt && make  top_distdir=../../../syslog-ng-4.3.1 distdir=../../../syslog-ng-4.3.1/lib/ivykis/test.mt \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
:
test -d "../../../syslog-ng-4.3.1/lib/ivykis/test.mt" || mkdir "../../../syslog-ng-4.3.1/lib/ivykis/test.mt"
test -n ":" \
|| find "../../../syslog-ng-4.3.1/lib/ivykis/test.mt" -type d ! -perm -755 \
        -exec chmod u+rwx,go+rx {} \; -o \
  ! -type d ! -perm -444 -links 1 -exec chmod a+r {} \; -o \
  ! -type d ! -perm -400 -exec chmod a+r {} \; -o \
  ! -type d ! -perm -444 -exec /bin/bash /source/lib/ivykis/test.mt/install-sh -c -m a+r {} {} \; \
|| chmod -R a+r "../../../syslog-ng-4.3.1/lib/ivykis/test.mt"
 (cd contrib && make  top_distdir=../../../syslog-ng-4.3.1 distdir=../../../syslog-ng-4.3.1/lib/ivykis/contrib \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
 (cd iv_getaddrinfo && make  top_distdir=../../../../syslog-ng-4.3.1 distdir=../../../../syslog-ng-4.3.1/lib/ivykis/contrib/iv_getaddrinfo \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
 (cd kojines && make  top_distdir=../../../../syslog-ng-4.3.1 distdir=../../../../syslog-ng-4.3.1/lib/ivykis/contrib/kojines \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
 (cd man3 && make  top_distdir=../../../syslog-ng-4.3.1 distdir=../../../syslog-ng-4.3.1/lib/ivykis/man3 \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
 (cd misc && make  top_distdir=../../../syslog-ng-4.3.1 distdir=../../../syslog-ng-4.3.1/lib/ivykis/misc \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make  distdir-am
test -n ":" \
|| find "../../syslog-ng-4.3.1/lib/ivykis" -type d ! -perm -755 \
        -exec chmod u+rwx,go+rx {} \; -o \
  ! -type d ! -perm -444 -links 1 -exec chmod a+r {} \; -o \
  ! -type d ! -perm -400 -exec chmod a+r {} \; -o \
  ! -type d ! -perm -444 -exec /bin/bash /source/lib/ivykis/install-sh -c -m a+r {} {} \; \
|| chmod -R a+r "../../syslog-ng-4.3.1/lib/ivykis"
test -n "" \
|| find "syslog-ng-4.3.1" -type d ! -perm -755 \
        -exec chmod u+rwx,go+rx {} \; -o \
  ! -type d ! -perm -444 -links 1 -exec chmod a+r {} \; -o \
  ! -type d ! -perm -400 -exec chmod a+r {} \; -o \
  ! -type d ! -perm -444 -exec /bin/bash /source/install-sh -c -m a+r {} {} \; \
|| chmod -R a+r "syslog-ng-4.3.1"
tardir=syslog-ng-4.3.1 && false | eval GZIP= gzip --best -c >syslog-ng-4.3.1.tar.gz
if test -d "syslog-ng-4.3.1"; then find "syslog-ng-4.3.1" -type d ! -perm -200 -exec chmod u+w {} ';' && rm -rf "syslog-ng-4.3.1" || { sleep 5 && rm -rf "syslog-ng-4.3.1"; }; else :; fi
(dbld)asantos@mymachine:/dbld/build$ tar -tzvf syslog-ng-4.3.1.tar.gz
(dbld)asantos@mymachine:/dbld/build$
```

Do you know what is wrong?
Thanks in advance, Alex

**Reply by amdrsantos** [reply]
@bazsi , @alltilla, can you give some support?

**Comment by czanik** [comment]
One possible solution is to compile everything with DBLD, just not using the syslog-ng-mod-mongodb sub-package:

- check out the syslog-ng source: ```git clone https://github.com/syslog-ng/syslog-ng/```
- change to the 8.3.1 release: ```git checkout 317472438cee5fab529dabe286fae650d764ac65```
- check the dbld options: ```dbld/rules```
- build packages for your choice of Debian: ```dbld/rules deb-debian-bullseye```

You can find the results under the ```dbld/build``` directory. Just make sure that you do not install the mongodb-related package.

**Comment by amdrsantos** [comment]
Thanks Peter,
But I have a customized module to compile. So I want to compile from the sources.

I do not understand why compiling from inside the docker does not work:
```
/dbld/bootstrap
cd /dbld/build
make -j
make install
make dist
```
while using: `dbld/rules deb-debian-bookworm` works.

**Reply by therandomstring** [reply]
You could also build your modified codebase with dbld, as it is not limited to the contents of the main git repo.
Also I do not understand why you insist on compiling syslog-ng manually from a dbld shell, if the deb rule works.

Could you please confirm that your changes compile with the deb rule?

**Reply by amdrsantos** [reply]
My bad.
Using: `dbld/rules deb-debian-bookworm` also does not work. (I had a tarball on the parent directory that was making it not to fail)

When I clean up everything and then recompile again:
```
git clone https://github.com/syslog-ng/syslog-ng/
cd syslog-ng
git checkout syslog-ng-4.3.1
dbld/rules deb-debian-bookworm
```

I get the same problem as compiling manually:
```
if test -d "syslog-ng-4.3.1"; then find "syslog-ng-4.3.1" -type d ! -perm -200 -exec chmod u+w {} ';' && rm -rf "syslog-ng-4.3.1" || { sleep 5 && rm -rf "syslog-ng-4.3.1"; }; else :; fi
Your tarball is in /build, also available on the host in $(top_srcdir)/dbld/build
Added new user: asantos
/dbld/pkg-tarball: line 15: cd: syslog-ng-4.3.1: No such file or directory
make: *** [dbld/rules:143: pkg-tarball-tarball] Error 1
```

Complete output:
[output.txt](https://github.com/user-attachments/files/16818524/output.txt)


The same problem is happening compiling manually or with deb rule.

**Reply by therandomstring** [reply]
Unfortunately I couldn't replicate your problem, as this same sequence of commands resulted in non-empty deb packages.

I would try to investigate for the reason the directory goes missing in the middle of the procedure. It likely gets created, as dbld should use it earlier in the process.

**Comment by amdrsantos** [comment]
I was able to compile it on a different machine.
Sorry for the inconvenience.
Thanks, @lmesz for your message: 

> Is this not by any chance a non-x86-64 environment, where the docker image is running virtualized through multiple layers of clever file system tricks? For example, I saw similar issues on Apple Silicon Macbooks with Docker Desktop.
